### ISSUE TITLE
namespaceSelector not working on RKE(rancher kubernetes engine) with CNI calico

### ISSUE THREAD

**Issue by mak241265** [projectcalico/calico#9402]
Hello everyone..i have seen a few post complaining about namespaceSelector in network policy.i still have this problem and can not figure it out.

I have installed redis from bitnami and inside of helm chart i have enabled network policy with some customization:

```
apiVersion: networking.k8s.io/v1
kind: NetworkPolicy
metadata:
  annotations:
    meta.helm.sh/release-name: redis-test
    meta.helm.sh/release-namespace: default
  creationTimestamp: "2024-10-28T12:21:16Z"
  generation: 1
  labels:
    app.kubernetes.io/instance: redis-test
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: redis
    helm.sh/chart: redis-17.9.5
  name: redis-test
  namespace: default
  resourceVersion: "32158854"
  uid: cce1065f-aed3-4312-aaaa-989be50591ab
spec:
  egress:
  - ports:
    - port: 53
      protocol: UDP
  - ports:
    - port: 6379
      protocol: TCP
    to:
    - podSelector:
        matchLabels:
          app.kubernetes.io/instance: redis-test
          app.kubernetes.io/name: redis
  ingress:
  - from:
    - podSelector:
        matchLabels:
          redis-test-client: "true"
    - podSelector:
        matchLabels:
          app.kubernetes.io/instance: redis-test
          app.kubernetes.io/name: redis
    - namespaceSelector:
        matchLabels:
          redis: external
      podSelector:
        matchLabels:
          redis-client: allow
    ports:
    - port: 6379
      protocol: TCP
  - ports:
    - port: 9121
      protocol: TCP
  podSelector:
    matchLabels:
      app.kubernetes.io/instance: redis-test
      app.kubernetes.io/name: redis
  policyTypes:
  - Ingress
  - Egress
status: {}
```

## Expected Behavior

from my understanding any pods which is in namespace which has a label called redis: external and also the pods its self with label redis-client: allow can connect to redis on port 6379

## Current Behavior

....i had conducted many tests and everything works well if the pod its self is inside the same namespace with redis

## Your Environment

My info enviornment  is below:
rke version: v1.4.6
k8s version: v1.26.4
calico version: v3.25.1
Os version: ubuntu 22.04

**Comment by mak241265** [comment]
It seems that if the redis client is on the same node as redis instance, then the connection can be set without probelm

**Comment by caseydavenport** [comment]
> then the connection can be set without problem but on different nodes then no connection.

This seems to suggest that the problem isn't so much with network policy, but with the cross-node networking configuration. A few potential things:

- It could just be the cross host networking is broken. However I'd expect this to be much more obvious in this case.
- It could be that NAT is being performed somewhere when communicating across hosts, which would cause the arriving packet to not be recognized by the network policy rules.

I suspect the latter is more likely - would be worth trying to do a packet capture, or verify that there is or is not NAT occurring when sending packets to redis when the client is on a different node.

**Comment by mak241265** [comment]
Actually when you use RKE to install k8s cluster you have an option to tell RKE to install CNI(calico or..) for you.when you choose this option most configuration are done by RKE.
one of the thing that is configured by default in RKE is kube_controller cluster_cidr which should be matched with CNI cidr. 
In our environment we told RKE not to install any CNI for us because we wanted to configure it manually and also we were not noticed that kube_controller cluster_cidr is set by default  by rke (cluster_cidr: 10.42.0.0/16) the we installed calico with different range and that's made conflict and faced some problem..so we defined new ippool and we set the pool as the same as cluster_cidr and then everything worked well.
